src/interface/state_descriptor.hpp

```diff
@@ -118,52 +191,55 @@ class StateDescriptor {
-  const Dictionary<Metadata> &AllFields() const { return metadataMap_; }
-  const Dictionary<std::unordered_map<int, Metadata>> &AllSparseFields() const {
-    return sparseMetadataMap_;
-  }
-  const Dictionary<Metadata> &AllSwarms() const { return swarmMetadataMap_; }
-  const Dictionary<Metadata> &AllSwarmValues(const std::string &swarm_name) const {
+  const auto &AllFields() const noexcept { return metadataMap_; }
+  const auto &AllSparsePools() const noexcept { return sparsePoolMap_; }
+  const auto &AllSwarms() const noexcept { return swarmMetadataMap_; }
+  const auto &AllSwarmValues(const std::string &swarm_name) const noexcept {
     return swarmValueMetadataMap_.at(swarm_name);
   }
-  bool FieldPresent(const std::string &field_name) const {
-    return metadataMap_.count(field_name) > 0;
+  bool FieldPresent(const std::string &base_name, int sparse_id = InvalidSparseID) const
+      noexcept {
+    return metadataMap_.count(VarID(base_name, sparse_id)) > 0;
   }
-  bool SparsePresent(const std::string &field_name) const {
-    return sparseMetadataMap_.count(field_name) > 0;
+  bool SparseBaseNamePresent(const std::string &base_name) const noexcept {
+    return sparsePoolMap_.count(base_name) > 0;
   }
-  bool SparsePresent(const std::string &field_name, int i) const {
-    if (sparseMetadataMap_.count(field_name) > 0) {
-      return sparseMetadataMap_.at(field_name).count(i) > 0;
-    }
-    return false;
-  }
-  bool SwarmPresent(const std::string &swarm_name) const {
+  bool SwarmPresent(const std::string &swarm_name) const noexcept {
     return swarmMetadataMap_.count(swarm_name) > 0;
   }
   bool SwarmValuePresent(const std::string &value_name,
-                         const std::string &swarm_name) const {
+                         const std::string &swarm_name) const noexcept {
     if (!SwarmPresent(swarm_name)) return false;
     return swarmValueMetadataMap_.at(swarm_name).count(value_name) > 0;
   }
 
   // retrieve metadata for a specific field
-  Metadata &FieldMetadata(const std::string &field_name) {
-    return metadataMap_[field_name];
-  }
-
-  Metadata &FieldMetadata(const std::string &field_name, int i) {
-    return sparseMetadataMap_[field_name][i];
+  const Metadata &FieldMetadata(const std::string &base_name,
+                                int sparse_id = InvalidSparseID) const {
+    const auto itr = metadataMap_.find(VarID(base_name, sparse_id));
+    PARTHENON_REQUIRE_THROWS(itr != metadataMap_.end(),
+                             "FieldMetadata: Non-existent field: " +
+                                 MakeVarLabel(base_name, sparse_id));
+    return itr->second;
+  }
+
+  const auto &GetSparsePool(const std::string &base_name) const noexcept {
+    static const SparsePool empty_pool("EmptySparsePool", Metadata({Metadata::Sparse}));
+
+    const auto itr = sparsePoolMap_.find(base_name);
+    if (itr == sparsePoolMap_.end()) {
+      return empty_pool;
+    } else {
+      return itr->second;
+    }
```

Comment:
Calling this `GetOrAddSparsePool` would not be accurate, because it doesn't actually add the pool. If you request a sparse pool "foobar" that doesn't exist, it returns a static, empty pool called "EmptySparsePool". Next time you request "foobar", you will still get "EmtpySparsePool", it won't add an empty pool called "foobar".

This is related to https://github.com/lanl/parthenon/pull/535#discussion_r645902176

The previous behavior `FieldMetadata` was to return a default-constructed metadata instance if the user asked for the metadata of a non-existent field. So I tried to mimic that behavior for sparse pools. However, @Yurlungur and I decided in that case that it would be preferable to throw, since there is a method `FieldPresent` to check if a particular field exists. In this case, I would actually also prefer to just throw if the user asks for a non-existent pool, since there is a function `SparseBaseNamePresent` that can be used to query if a particular pool exists or not.

Any objections to making this throw if the requested pool doesn't exist?

Reply:
No objections here. I'm also in favor of explicit code.